fdbserver/ClusterController.actor.cpp

```diff
@@ -1691,6 +1700,196 @@ class ClusterControllerData {
 		return idUsed;
 	}
 
+	std::unordered_set<NetworkAddress> getServersWithLinkProblem() {
+		if (!SERVER_KNOBS->ENABLE_PEER_HEALTH_MONITOR) {
+			return {};
+		}
+
+		std::unordered_map<NetworkAddress, int> degradationCounts;
+		for (const auto& [server, degradedRemotePeers] : degradedPeers) {
+			for (const auto& [degradedServer, startTime] : degradedRemotePeers) {
+				if (now() - startTime < SERVER_KNOBS->PEER_DEGRADATION_INTERVAL) {
+					continue;
+				}
+				degradationCounts[degradedServer]++;
+			}
+		}
+
+		std::unordered_set<NetworkAddress> excludeServers;
+		for (const auto& [server, degradationCount] : degradationCounts) {
+			if (degradationCount <= SERVER_KNOBS->MAX_PEER_DEGRADATION_TO_EXCLUDE) {
+				excludeServers.insert(server);
+				continue;
+			}
+		}
+		return excludeServers;
+	}
+
+	void updatePeerHealth(const UpdatePeerHealthRequest& req) {
+		double currentTime = now();
+		if (req.degradedPeers.empty()) {
```

Comment:
It's possible a node's peers all become healthy at the same time, we need to remove them from degradedPeers